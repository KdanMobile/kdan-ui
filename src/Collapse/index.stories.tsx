import React from "react";

import Button from "../Button";
import Collapse, { CollapseProps } from "./";
import Box from "../Box";
import { Container } from "../component/Decorator";

export default {
  title: "Collapse",
  component: Collapse,
  decorators: [
    (story: () => React.ReactNode): React.ReactNode => (
      <Container padding>{story()}</Container>
    ),
  ],
  parameters: {
    docs: {
      description: {
        component: "展開項目",
      },
    },
  },
};

export const Basic = ({}: CollapseProps): React.ReactNode => {
  return (
    <Box maxWidth="200px">
      <Collapse trigger={<Button>Click</Button>}>
        <p style={{ backgroundColor: "white", height: "200px" }}>
          inside content
        </p>
      </Collapse>
      <p>outside content</p>
    </Box>
  );
};

Basic.args = {} as CollapseProps;

export const PassFunctionToTrigger = ({}: CollapseProps): React.ReactNode => (
  <Box maxWidth="200px">
    <Collapse
      trigger={(open: boolean) => (
        <Button>Click {open ? "open" : "close"}</Button>
      )}>
      <p style={{ backgroundColor: "white", height: "200px" }}>content</p>
    </Collapse>
  </Box>
);

PassFunctionToTrigger.args = {} as CollapseProps;
