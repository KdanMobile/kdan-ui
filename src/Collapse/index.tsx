import React, { useRef, useState } from "react";

import { Wrapper, TriggerWrapper, ContentWrapper, Inner } from "./styled";

export type CollapseProps = {
  trigger: React.ReactNode;
  children: React.ReactNode;
};

const Collapse: React.FC<CollapseProps> = ({
  trigger,
  children,
}: CollapseProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const [open, setOpen] = useState(false);

  const handleClick = () => {
    setOpen((current) => !current);
  };

  return (
    <Wrapper>
      <TriggerWrapper onClick={handleClick}>
        {typeof trigger === "function" ? trigger(open) : trigger}
      </TriggerWrapper>
      <ContentWrapper
        style={{ height: open ? `${innerRef.current?.clientHeight}px` : 0 }}>
        <Inner ref={innerRef}>{children}</Inner>
      </ContentWrapper>
    </Wrapper>
  );
};

export default Collapse;
