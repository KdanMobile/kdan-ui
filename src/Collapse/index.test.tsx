import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import "jest-styled-components";
import "@testing-library/jest-dom";

import Collapse from "./";

describe("Collapse", () => {
  afterEach(cleanup);

  test("open collapse", () => {
    const { container, getByText } = render(
      <Collapse trigger={<button>click</button>} defaultOpen>
        content
      </Collapse>
    );

    fireEvent.click(getByText("click"));

    expect(container).toMatchSnapshot();
  });
});
