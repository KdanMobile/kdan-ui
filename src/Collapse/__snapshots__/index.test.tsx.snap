// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse open collapse 1`] = `
.c0 {
  color: inherit;
  overflow: hidden;
  -webkit-transition: height 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: height 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  height: 0;
}

<div>
  <div>
    <div>
      <button>
        click
      </button>
    </div>
    <div
      class="c0"
      style="height: 0px;"
    >
      <div
        class=""
      >
        content
      </div>
    </div>
  </div>
</div>
`;
